restapi/src/test/java/io/stargate/web/docsapi/service/query/ExpressionParserIntTest.java

```diff
@@ -131,6 +136,124 @@ public void andWrappingRespected() {
                                     }));
               });
     }
+
+    @Test
+    public void complex() {
+      // a & (b | (c & d))
+      ObjectNode root = mapper.createObjectNode();
+      root.set("a", mapper.createObjectNode().put("$eq", "a"));
+      root.set(
+          "$or",
+          mapper
+              .createArrayNode()
+              .add(mapper.createObjectNode().set("b", mapper.createObjectNode().put("$eq", "b")))
+              .add(
+                  mapper
+                      .createObjectNode()
+                      .set(
+                          "$and",
+                          mapper
+                              .createArrayNode()
+                              .add(
+                                  mapper
+                                      .createObjectNode()
+                                      .set("c", mapper.createObjectNode().put("$eq", "c")))
+                              .add(
+                                  mapper
+                                      .createObjectNode()
+                                      .set("d", mapper.createObjectNode().put("$eq", "d"))))));
+
+      Expression<FilterExpression> result =
+          service.constructFilterExpression(Collections.emptyList(), root, false);
+
+      assertThat(result)
+          .isInstanceOfSatisfying(
+              And.class,
+              a -> {
+                List<?> children = a.getChildren();
+                assertThat(children)
+                    .hasSize(2)
+                    .anySatisfy(
+                        c ->
+                            assertThat(c)
+                                .isInstanceOfSatisfying(
+                                    FilterExpression.class,
+                                    first -> {
+                                      assertThat(first.getOrderIndex()).isZero();
+                                      assertThat(first.getFilterPath().getField()).isEqualTo("a");
+                                    }))
+                    .anySatisfy(
+                        c ->
+                            assertThat(c)
+                                .isInstanceOfSatisfying(
+                                    Or.class,
+                                    or -> {
+                                      List<?> orChildren = or.getChildren();
+                                      assertThat(orChildren)
+                                          .hasSize(2)
```

Comment:
WDYT about using `toString()` of the filter expression for testing? It does not have to be a well-defined representation for end-users to work with, just stable enough for asserting in test. I believe it would make test code more readable.

Reply:
Nope, this can hide too many stuff.. Keeping it like this for now, it's not readable due to the formatting now the actual code.